### ISSUE TITLE
Regression with quotations of List.sum on non-primitive type

### ISSUE THREAD

**Issue by dsyme** [dotnet/fsharp#10364]
### Description

When using `--langversion:5.0` (the default in F# 5.0), taking the quotation of a use of `List.sum` on a non-primitive type causes a compilation exception related to generating the witness for the `Zero` property.

This will apply to a quotation of any code calling `let inline` SRTP code which uses a constraint on a `get_Foo` property method or any SRTP constraint that takes no arguments `member M : unit -> ...` 

### Repro

```fsharp
type Point =
    { x: int; y: int }
    static member Zero = { x=0; y=0 }
    static member (+) (p1, p2) = { x= p1.x + p2.x; y = p1.y + p2.y }
let p1 = {x=1; y=10}

<@ List.sum [p1] @>
```

gives
```
stdin(0,1): error FS0073: internal error: The lists had different lengths.
list2 is 1 element shorter than list1
Parameter name: list2 (ArgumentException)
Stack Trace
FSharp.Compiler.ErrorLogger+InternalError: The lists had different lengths.
list2 is 1 element shorter than list1
Parameter name: list2 (ArgumentException)stdin (0,0--0,0) IsSynthetic=false
```

Another repro is 
```fsharp
#r "System.Numerics.Vectors.dll";;
<@ List.sum [System.Numerics.Vector4.Zero] @>;;
```

### Workarounds

One workaround is to use `--langversion:4.7`

Another possible workaround is to define a non-inlined helper function outside the quotation, e.g. as follows.  If necessary label the function as `ReflectedDefinition`. This may not work if the quotation is being processed requires access to the full quotation tree.

```fsharp
type Point =
    { x: int; y: int }
    static member Zero = { x=0; y=0 }
    static member (+) (p1, p2) = { x= p1.x + p2.x; y = p1.y + p2.y }
let p1 = {x=1; y=10}

let sumPoints (points: Point list) = List.sum points

<@ sumPoints [p1] @>
```

### Severity

This bug has not yet been seen in actual user code, but was instead noticed as part of https://github.com/dotnet/fsharp/pull/9510, which applies witness generation to a much broader range of code (all code going through Fable).

However it's likely to affect some user code.

Since a workaround is to use `--langversion:4.7` it is likely that the actual severity in practice will be low.

**Comment by dsyme** [comment]
The relevant fix is here: https://github.com/dotnet/fsharp/pull/9510/commits/316af3b748f59a5d73f0fd5a7f9e53bb030b69ca#diff-6c2a9af7af19c0b607dfef044bea61f4cea886c9c9a867c54a68dd105254e722R1875

We can take it out of that PR and isolate it in a new PR if necessary

**Comment by cartermp** [comment]
@dsyme at most we're looking at a servicing release for .NET  5 or 16.8. Is that necessary, or would you be fine just having this as a known issue and letting the fix flow into 16.9 and the patch to .NET 5 that will go with it? Both are set to go out around February.

**Comment by dsyme** [comment]
> @dsyme at most we're looking at a servicing release for .NET 5 or 16.8. Is that necessary, or would you be fine just having this as a known issue and letting the fix flow into 16.9 and the patch to .NET 5 that will go with it? Both are set to go out around February.

I presume the workaround will be sufficient for customers, I'll let you make the call, and how to document it as a known issue so it's discoverable by a search.

That is, assuming the problem actually gets reported in practice.  It feels like it will but I may be wrong

**Comment by cartermp** [comment]
Sounds goood, i'll tag it as 16.9.

**Comment by dsyme** [comment]
This was fixed in #9510
